nessie/lakectl_util_test.go

```diff
@@ -0,0 +1,233 @@
+	runCmdAndVerifyWithFile(t, cmd, false, isTerminal, goldenFile, vars)
+}
+
+func runCmdAndVerifyFailureWithFile(t *testing.T, cmd string, isTerminal bool, goldenFile string, vars map[string]string) {
+	runCmdAndVerifyWithFile(t, cmd, true, isTerminal, goldenFile, vars)
+}
+
+func runCmdAndVerifyWithFile(t *testing.T, cmd string, expectFail bool, isTerminal bool, goldenFile string, vars map[string]string) {
+	goldenFile = "golden/" + goldenFile + ".golden"
+
+	if *update {
+		result, _ := runShellCommand(cmd, isTerminal)
+		s, err := embedVariables(string(result), vars)
+		require.NoError(t, err, "Variable embed failed - %s", err)
+		err = ioutil.WriteFile(goldenFile, []byte(sanitize(s, vars)), 0777)
+		require.NoError(t, err, "Failed to write file %s", goldenFile)
+		return
+	}
+	content, err := ioutil.ReadFile(goldenFile)
+	if err != nil {
+		t.Fatal("Failed to read ", goldenFile, err)
+	}
+	expected := sanitize(string(content), vars)
+	runCmdAndVerifyResult(t, cmd, expectFail, isTerminal, expected, vars)
+}
+
+func runCmdAndVerifySuccess(t *testing.T, cmd string, isTerminal bool, expected string, vars map[string]string) {
+	runCmdAndVerifyResult(t, cmd, false, isTerminal, expected, vars)
+}
+
+func runCmdAndVerifyFailure(t *testing.T, cmd string, isTerminal bool, expected string, vars map[string]string) {
+	runCmdAndVerifyResult(t, cmd, true, isTerminal, expected, vars)
+}
+
+func runCmdAndVerifyResult(t *testing.T, cmd string, expectFail bool, isTerminal bool, expected string, vars map[string]string) {
+	expanded, err := expandVariables(expected, vars)
+	if err != nil {
+		t.Fatal("Failed to extract variables for", cmd)
+	}
+	result, err := runShellCommand(cmd, isTerminal)
+	if expectFail {
+		require.Error(t, err, "Expected error in %s command did not occur. Output -%s", cmd, string(result))
+	} else {
+		require.NoError(t, err, "Failed to run %s command - %s", cmd, string(result))
+	}
+	require.Equal(t, expanded, sanitize(string(result), vars), "Unexpected output for %s command", cmd)
+}
+
+var (
+	timeStampRegexp = regexp.MustCompile(`timestamp: \d+\r?\n`)
+	timeRegexp      = regexp.MustCompile(`\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} [\+|-]\d{4} \w{1,3}`)
+	commitIdRegExp  = regexp.MustCompile(`[\d|a-f]{64}`)
+)
+
+func removeProgramTimestamp(output string) string {
+	s := timeStampRegexp.ReplaceAll([]byte(output), []byte("timestamp: <TIMESTAMP>"))
+	s = timeRegexp.ReplaceAll(s, []byte("<DATE> <TIME> <TZ>"))
+	return string(s)
+}
+
+func removeRandomObjectKey(output string, objectPrefix string) string {
+	objectKeyRegExp := regexp.MustCompile(objectPrefix + `/[\d|a-f]{32}`)
+	s := objectKeyRegExp.ReplaceAll([]byte(output), []byte(objectPrefix+"/<OBJECT_KEY>"))
+	return string(s)
+}
+
+func removeCommitID(output string) string {
+	s := commitIdRegExp.ReplaceAll([]byte(output), []byte("<COMMIT_ID>"))
+	return string(s)
+}
+
+//
+// Utilities tests
+//
```

Comment:
Initially, that was the way I did it, but I got linter complaining about private functions not being used, so based on Barak's advice, I moved the functions the a file named *_test.go. Now, since this file already had the tests for these finctions I left it that way.
These are just tests to protect the utility functions, and not the lakectl actual tests, so I can live with that, but if you think it is a bad practice/unclean approach - I can move the tests out.
WDYT?

Reply:
We have numerous `testutil` files in lakeFS, perhaps draw inspiration from those?  (If you put them into a separate directory, then they become a separate package.  Now you can export them and test them just as you would any other function!)

Reply:
Done